recipes/PeoplesSpeech/ASR/transformer/train.py

```diff
@@ -0,0 +1,458 @@
+    # name for the sampler already, also it's not an id, but an audio_path.
+    train_data = ds[0].rename_column("id", "audio_id")
+    valid_data = ds[1].rename_column("id", "audio_id")
+    test_data = ds[2].rename_column("id", "audio_id")
+
+    # We need to get the full list of durations of all samples to enable
+    # bucketing from the dynamic batch sampler. We do it that way instead
+    # of the usual iterable because the HF dataset ALWAYS open the file
+    # when called, which means that the dynamic sampling needs to read the
+    # 1.5M audio samples from disk.... using a list instead is much master.
+    train_len_list = list(
+        train_data.select_columns("duration_ms")["duration_ms"]
+    )
+    val_len_list = list(valid_data.select_columns("duration_ms")["duration_ms"])
+
+    train_data = sb.dataio.dataset.DynamicItemDataset.from_arrow_dataset(
+        train_data,
+    )
+
+    valid_data = sb.dataio.dataset.DynamicItemDataset.from_arrow_dataset(
+        valid_data,
+    )
+
+    test_data = sb.dataio.dataset.DynamicItemDataset.from_arrow_dataset(
+        test_data,
+    )
+
+    datasets = [train_data, valid_data, test_data]
+
+    # 2. Define audio pipeline:
+    @sb.utils.data_pipeline.takes("audio")
+    @sb.utils.data_pipeline.provides("sig")
+    def audio_pipeline(audio):
+        if audio["sampling_rate"] != 16000:
+            sig = torchaudio.transforms.Resample(
+                audio["sampling_rate"],
+                16000,
+            )(audio["array"])
+        else:
+            sig = audio["array"].astype(np.single)
+        yield sig
+
+    sb.dataio.dataset.add_dynamic_item(datasets, audio_pipeline)
+
+    # 3. Define text pipeline:
+    @sb.utils.data_pipeline.takes("text")
+    @sb.utils.data_pipeline.provides(
+        "wrd", "tokens_list", "tokens_bos", "tokens_eos", "tokens"
+    )
+    def text_pipeline(wrd):
+
+        wrd = english_specific_preprocess(wrd)
+        yield wrd
+        tokens_list = tokenizer.sp.encode_as_ids(wrd)
+        yield tokens_list
+        tokens_bos = torch.LongTensor([hparams["bos_index"]] + (tokens_list))
+        yield tokens_bos
+        tokens_eos = torch.LongTensor(tokens_list + [hparams["eos_index"]])
+        yield tokens_eos
+        tokens = torch.LongTensor(tokens_list)
+        yield tokens
+
+    sb.dataio.dataset.add_dynamic_item(datasets, text_pipeline)
+
+    # 4. Set output:
+    sb.dataio.dataset.set_output_keys(
+        datasets,
+        ["id", "audio_id", "sig", "wrd", "tokens_bos", "tokens_eos", "tokens"],
+    )
+
+    # 5. If Dynamic Batching is used, we instantiate the needed samplers.
+    train_batch_sampler = None
+    valid_batch_sampler = None
+
+    dynamic_hparams = hparams["dynamic_batch_sampler"]
+    dynamic_hparams_valid = hparams["dynamic_batch_sampler_valid"]
+    num_buckets = dynamic_hparams["num_buckets"]
+
+    train_batch_sampler = DynamicBatchSampler(
+        train_data,
+        dynamic_hparams["max_batch_length"] * 1000,
```

Comment:
are you converting the units to seconds or something like that? maybe you should create a proper variable to better understand what are you doing here

Reply:
Yeah, I had put a comment at some point ... but it disapeared.